### DISCUSSION TITLE
Configuring sensor update/refresh time?

### DISCUSSION THREAD

**Discussion by AlexK98** [make-all/tuya-local#2654] [Q&A]
Is there sensor update/refresh time built-in somewhere in Tuya Local? Can it be configurable?

I have T34-Smart Plug+ (wifi plug) and if I do not open Smart Life app, the chart shows update every minute or so, but when Smart Life app is open the update time is 2 seconds. Close Smart Life app and update goes back to 1 minute cycle.

**Comment by make-all** [comment]
This integration is not polling, it is push, and the updates should come as soon as the device sends them.  The behaviour you describe suggests that there is a dp that will manually prompt the device into sending some data. For some devices, any writable dp will do, so if the device has a timer, setting the timer to 0 periodically might work without causing too much side effect.
For other devices, there is a specific dp to use for this, and where known it has been added to the config as a "Refresh" button, which you can press with an automation.

**Reply by AlexK98** [reply]
Only 1 of 11 devices I own has this "Refresh" DP. In Smart Life app, for this device there is no corresponding button for "refresh" DP. It was probably meant for internal use, so that when user opens device's card in Tuya app it will start pushing "refresh" or something. I don't know.

Basically, I have no idea how updating is done in your integration or any other Tuya integrations - polling or pushing.
I'm just asking if we can have an option to configure the poll/push period per device or in general?

ps. along with you nice integration I found **localtuya** fork by [xZetsubou](https://github.com/xZetsubou/hass-localtuya/). And it has an option for sort of "refresh interval" per device. But the problem is that it does not handle base64 values, like your integration.

**Reply by stephanmiehe** [reply]
@make-all / @AlexK98 did either of you find a path? I have the following model:

```yaml
name: Energy monitoring smartplug
products:
  - id: jdj6ccklup7btq3a
    manufacturer: Tongou
    model: TO-Q-SY1-JWT
```

I have the same issue. Updates do push but it's rather slow so the watts used isn't as accurate as I'd like. If I open the Tuya app and open the device, suddenly I'm getting refreshed data every few seconds vs. 30-60 seconds. Even after closing the app, that refresh interval lasts for a while but eventually returns to the slower one

Any ideas how to figure out what happens so I could manually replicate it?

Specification and properties of the device:

```json
{
  "result": {
    "category": "dlq",
    "functions": [
      {
        "code": "switch",
        "desc": "{}",
        "name": "开关",
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "countdown_1",
        "desc": "{\"unit\":\"s\",\"min\":0,\"max\":86400,\"scale\":0,\"step\":1}",
        "name": "开关倒计时",
        "type": "Integer",
        "values": "{\"unit\":\"s\",\"min\":0,\"max\":86400,\"scale\":0,\"step\":1}"
      },
      {
        "code": "relay_status",
        "desc": "{\"range\":[\"power_off\",\"power_on\",\"last\"]}",
        "name": "上电状态设置",
        "type": "Enum",
        "values": "{\"range\":[\"power_off\",\"power_on\",\"last\"]}"
      },
      {
        "code": "light_mode",
        "desc": "{\"range\":[\"relay\",\"pos\",\"none\",\"on\"]}",
        "name": "指示灯状态设置",
        "type": "Enum",
        "values": "{\"range\":[\"relay\",\"pos\",\"none\",\"on\"]}"
      },
      {
        "code": "child_lock",
        "desc": "{}",
        "name": "童锁开关",
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "cycle_time",
        "desc": "{\"maxlen\":255}",
        "name": "循环定时",
        "type": "String",
        "values": "{\"maxlen\":255}"
      },
      {
        "code": "online_state",
        "desc": "{\"range\":[\"online\",\"offline\"]}",
        "name": "查询状态",
        "type": "Enum",
        "values": "{\"range\":[\"online\",\"offline\"]}"
      }
    ],
    "status": [
      {
        "code": "switch",
        "name": "开关",
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "countdown_1",
        "name": "开关倒计时",
        "type": "Integer",
        "values": "{\"unit\":\"s\",\"min\":0,\"max\":86400,\"scale\":0,\"step\":1}"
      },
      {
        "code": "add_ele",
        "name": "增加电量",
        "type": "Integer",
        "values": "{\"unit\":\"\",\"min\":0,\"max\":9999999,\"scale\":3,\"step\":100}"
      },
      {
        "code": "cur_current",
        "name": "当前电流",
        "type": "Integer",
        "values": "{\"unit\":\"mA\",\"min\":0,\"max\":100000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "cur_power",
        "name": "当前功率",
        "type": "Integer",
        "values": "{\"unit\":\"W\",\"min\":0,\"max\":999999,\"scale\":1,\"step\":1}"
      },
      {
        "code": "cur_voltage",
        "name": "当前电压",
        "type": "Integer",
        "values": "{\"unit\":\"V\",\"min\":0,\"max\":5000,\"scale\":1,\"step\":1}"
      },
      {
        "code": "test_bit",
        "name": "产测结果位",
        "type": "Integer",
        "values": "{\"min\":0,\"max\":5,\"scale\":0,\"step\":1}"
      },
      {
        "code": "voltage_coe",
        "name": "电压校准系数",
        "type": "Integer",
        "values": "{\"min\":0,\"max\":1000000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "electric_coe",
        "name": "电流校准系数",
        "type": "Integer",
        "values": "{\"min\":0,\"max\":1000000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "power_coe",
        "name": "功率校准系数",
        "type": "Integer",
        "values": "{\"min\":0,\"max\":1000000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "electricity_coe",
        "name": "电量校准系数",
        "type": "Integer",
        "values": "{\"min\":0,\"max\":1000000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "fault",
        "name": "故障告警",
        "type": "Bitmap",
        "values": "{\"label\":[\"ov_cr\",\"ov_vol\",\"ov_pwr\",\"ls_cr\",\"ls_vol\",\"ls_pow\"]}"
      },
      {
        "code": "relay_status",
        "name": "上电状态设置",
        "type": "Enum",
        "values": "{\"range\":[\"power_off\",\"power_on\",\"last\"]}"
      },
      {
        "code": "light_mode",
        "name": "指示灯状态设置",
        "type": "Enum",
        "values": "{\"range\":[\"relay\",\"pos\",\"none\",\"on\"]}"
      },
      {
        "code": "child_lock",
        "name": "童锁开关",
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "cycle_time",
        "name": "循环定时",
        "type": "String",
        "values": "{\"maxlen\":255}"
      },
      {
        "code": "online_state",
        "name": "online_state",
        "type": "Enum",
        "values": "{\"range\":[\"online\",\"offline\"]}"
      }
    ]
  },
```

**Reply by AlexK98** [reply]
@stephanmiehe
I did not find a way to get short refresh interval. Most of such devices expose 'Total *' DPs, and as they calculate this internally I decided it is enough for me.

**Reply by stephanmiehe** [reply]
Got it. Did you by chance man-in-the-middle the device to see what is happening between the device and the mobile app? I'm about to set that up in a lab environment but won't bother if that has already been tried

**Reply by make-all** [reply]
dp 66 which is exposed as a connectivity sensor is writable, and can be used to control the refresh rate. See commit b707f5f

This was not known when the config was initially written, but there have been several other devices with the same pattern now, so we now understand how this works. The dp is exposed as a button, when pressed it starts high-frequency updates and about 10 minutes later the device resets it, so if you want to monitor at high frequency continuously then you need an automation to periodically press the button.

We do not know why the manufacturer designed it this way, so there may be side effects from continuously using the high frequency mode (for example the device overheating, or the Tuya cloud account being frozen for excessive data use).
